### ISSUE TITLE
Flickering of the disparity image

### ISSUE THREAD

**Issue by TessaLin** [gengshan-y/high-res-stereo#44]
Using 20 frames image at same position, and  disparity image will flicker

photo and video:
https://drive.google.com/file/d/1P0hc1h0rokUyQbUM063c_SlsCZFn48P8/view?usp=sharing

**Comment by TessaLin** [comment]
Parameter:
kitti.tar
testres=0.5
max_disp=1280
level=1

**Comment by gengshan-y** [comment]
The pixel intensity changing over time, and the KITTI model being not very robust may cause the flickering.
Switching to Middlebury model with testres=3 (based on your image size) should give better results.

**Comment by TessaLin** [comment]
Change to Middlebury model,still have filckering issue
Parameter:
final-768px.tar
testres=3
max_disp=480
level=1

https://drive.google.com/file/d/1jdjg526eSbH55_GX7IKB59mEiwc_fdVS/view?usp=sharing

**Comment by gengshan-y** [comment]
Max disparity to search over seems small. Can you try increasing max_disp to 1280? If memory turns out to be an issue, you could set testres=2.

The flicking could be removed by passing --clean 1.0, which filters uncertain disparity predictions.

**Comment by daalberti1** [comment]
Hi, I think I'm dealing with the same problem. @TessaLin did you eventually find a solution?

I'm performing a 3D reconstruction based on the images (with resolution 1920*1080) that I receive from the binocular endoscope of the da Vinci Research Kit and actually the disparity map that I'm obtaining with this method seems to be quite good. However I noticed that the disparity map changes even if the endoscope stays still and as a consequence the corresponding point cloud that I compute through reprojectImageTo3D (by OpenCV) is not stable.

I followed @gengshan-y suggestions and then I supposed that the problem may have been related to the presence of gaussian noise in the images, so I tried to apply some filters but without any significant improvement.

**Comment by gengshan-y** [comment]
@elDvd96 If the problem is still there, can you share some example outputs? Ideally the model should be robust to small input noise because of training time color augmentations. Depending on how much noise are there in your inputs, one solution is to add gaussian noise to input images [here](https://github.com/gengshan-y/high-res-stereo/blob/cb60fde567066f59caea709dca1ac26b2ea27380/dataloader/MiddleburyLoader.py#L68) and finetune the model.

**Comment by daalberti1** [comment]
@gengshan-y Thank you for the reply and forgive me for the delay. Here you can find some examples of the rectified [input images](https://drive.google.com/drive/folders/1b44_yAbjdlBabgEj_kAuyDfNXsuUkbe0?usp=sharing) taken from the endoscope and the computed [disparity maps](https://drive.google.com/drive/folders/1YZsgQxoqrtE3nPjwkg_GrYc203m9-h13?usp=sharing) and [point clouds](https://drive.google.com/drive/folders/1jKvtHFrMRdZ9AzY8GunSHBoBtvPCp8gv?usp=sharing).
I'm also adding a [video](https://drive.google.com/drive/folders/1SOvjAWktbsDM3hMCq4QsUPRzTr60dQH0?usp=sharing) which show more clearly the problem.
I tried to tune the model following your suggestions but I didn't get better results.

**Comment by gengshan-y** [comment]
Hi @elDvd96, I looked at your outputs. The flickering in frame 0-5 seems expected, which is due to sensor noise (ISO too large?). To reduce sensor noise, you may want to use a lower ISO and longer exposure, but I'm not familiar with the one you are using.

The sudden change from frame 5 to frame 6 is abnormal, and seem to be a bug in saving disparity values, which can be fixed.

**Comment by daalberti1** [comment]
Hi @gengshan-y, unfortunately I think it's not possible to adjust ISO and exposure of the endoscope, anyway thank you for your suggestions, I'll look at least for the bug in saving disparity values.

**Comment by chenghao-matrix** [comment]
@gengshan-y 作者你好，感谢你的工作。我在使用sceneflow训练时出了问题：定位在这行代码：out = self.pool_convs[i](out)。意思是说，经过pool_convs这步处理过程，输入是[1,32,1,1,1]。而pool_convs本质是sepConv3d。又：sepConv3d中又包含nn.Conv3d和nn.BatchNorm3d。报错提示是：在训练模式下，[1,32,1,1,1]不能BN归一化处理。原始异常如下：{ValueError}ValueError('Expected more than 1 value per channel when training, got input size torch.Size([1, 32, 1, 1, 1])')

原始部分代码如下：
`class decoderBlock(nn.Module):
    """ 解码器网络 decoder """
    def forward(self, fvl):
        """
        :param fvl: 当前尺度匹配代价体 feat6==[1,32,D=6,8,12]
        """
        # --------------------- {conv3d,(C,C,1)} × 6
        # (1) 特征体通过6个3D卷积块进行滤波 Conv3d Block->left
        fvl = self.convs(fvl)  # [1,32,D=6,8,12]

        # --------------------- VPP：Pyramid Pooling
        # (2) 体积金字塔池化，为了高分辨率的输入，捕捉丰富的上下文信息
        if self.pool:
            fvl_out = fvl
            _, _, d, h, w = fvl.shape  # [1,32,D=6,8,12]

            # 生成4个等间距的分割因子pool_size。起始：1，终值：3，[1 1 2 3]
            for i, pool_size in enumerate(np.linspace(1, min(d, h, w) // 2, 4, dtype=int)):
                # 分析：分割因子pool_size增大，核变小，池化特征更粗略。size=(d,h,w)(2,18,24)表征将整个原图直接池化！！！
                kernel_size = (int(d / pool_size), int(h / pool_size), int(w / pool_size))  # (d,h,w)

                # 3D平均池化：(D,H,W)直接变(1,1,1)
                #           [1,32,D=2,H=18,W=24]->[1,32,1,1,1]
                out = F.avg_pool3d(fvl, kernel_size, stride=kernel_size)
                # 特征变换：(1x1x1卷积)进一步3d卷积处理池化特征
                #          [1,32,1,1,1]->[1,32,1,1,1]
                out = self.pool_convs[i](out)  # ？？？？？？？？？？？？？？？？？？？？？？？？？？？？报错。

                # 上采样：使用三线性插值将特征还原到原始特征的尺寸，确保与原始特征对齐，便于后续融合
                # [1,32,1,1,1]->[1,32,2,18,24]
                out = F.upsample(out, size=(d, h, w), mode='trilinear')`
